pkg/storage/fs/ocis/grants.go

```diff
@@ -0,0 +1,380 @@
+package ocis
+
+import (
+	"context"
+	"encoding/csv"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	userpb "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/pkg/appctx"
+	"github.com/cs3org/reva/pkg/errtypes"
+	"github.com/pkg/xattr"
+)
+
+func (fs *ocisfs) AddGrant(ctx context.Context, ref *provider.Reference, g *provider.Grant) (err error) {
+	log := appctx.GetLogger(ctx)
+	log.Debug().Interface("ref", ref).Interface("grant", g).Msg("AddGrant()")
+	var node *NodeInfo
+	if node, err = fs.pw.NodeFromResource(ctx, ref); err != nil {
+		return
+	}
+	if !node.Exists {
+		err = errtypes.NotFound(filepath.Join(node.ParentID, node.Name))
+		return
+	}
+
+	e, err := fs.getACE(g)
+	if err != nil {
+		return err
+	}
+
+	var attr string
+	if g.Grantee.Type == provider.GranteeType_GRANTEE_TYPE_GROUP {
+		attr = sharePrefix + "g:" + e.Principal
+	} else {
+		attr = sharePrefix + "u:" + e.Principal
+	}
+
+	np := filepath.Join(fs.pw.Root(), "nodes", node.ID)
+	if err := xattr.Set(np, attr, getValue(e)); err != nil {
```

Comment:
no, because a grant for the same grantee (principal / user / group) overrides the permissions of a previously existing grant.

In theory we could allow multiple grants for the same user, eg. with different expiry dates ... in theory ...